Add Support For Comparing Images

danieldresser-ie opened the pull request:
We've already talked and looked at this a fair bit, so hopefully this is about what you're expecting.

Things I'm aware of for discussion:

* Is the modification to WidgetAlgo:joinEdges worth it for the other clean ups it allows?
* Is D really a good hotkey when it means "disable" in so much of Gaffer's other UI?
* Is the opacity of the wipe tool right?  I don't think I can really go much thinner and have it smooth, so adding some transparency does seem to help it not feel too obtrusive.  One downside of the transparency is that the nice anti-aliased line no longer covers over the un-anti-aliased transition between the two images - this is only visible when the two images being compared are completely different colors though, which doesn't seem like a particularly important case.
* I've fixed/possibly-fixed the other issues with setting up the adjoined buttons, but I've still got the commit where I stomp over the border colors in the stylesheet.  Not sure what to do other than making this UI a special case.

johnhaddon commented on python/GafferUI/WidgetAlgo.py:
```diff
@@ -65,14 +65,19 @@ def joinEdges( widgets, orientation = None ) :
-			qtWidget = widget._qtWidget()
+		qtWidget = widget._qtWidget()
+
+		# Special case - many Gaffer Widgets place a single Widget inside a QGridLayout
+		# with a single item - in this case we need to apply the rounding to
+		# the internal widget.  Searching for the internal widget feels a little hacky,
+		# but it's more general than only supporting BoolPlugValueWidget, which is what
+		# we were doing before
+		## \todo Is there a better approach here, perhaps
+		# using the stylesheet?
+		if type( qtWidget ) == QtWidgets.QWidget and qtWidget.layout():
+			layoutItem = qtWidget.layout().itemAt( 0 )
+			if layoutItem  and layoutItem.widget():
+				qtWidget = layoutItem.widget()
```
It seemed like you might not have been completely happy with this heuristic, so I had a stab at my own : https://github.com/johnhaddon/gaffer/commit/5b37fdf43514ab516b376d4cf110b7481e1862ba. I think maybe that feels slightly more general?

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1063,11 +1107,152 @@ def __stateChanged( self, imageGadget ) :
+		self.getPlug().setValue( value )
+
+		node = self.getPlug().node()
+		inputImage = node._getPreprocessor()["in"]
```
Probably we should disable the comparison modes and the wipe tool in the ShaderView anyway?

danieldresser-ie replied:
Once we support  comparing to an arbitrary image node, I guess it could be useful to compare a shader ball to a reference image of some sort?  For the moment, I've just fixed this issue by using a better source plug.

johnhaddon reviewed (changes requested):
Thanks Daniel, this is nice!

> Is the modification to WidgetAlgo:joinEdges worth it for the other clean ups it allows?
   
I think so, yes - we'll want to use it more in future, so the more general it can be, the better. I had a stab at a slightly more general version of your improvement as well, in case it makes you feel any more positive about it.

> Is D really a good hotkey when it means "disable" in so much of Gaffer's other UI?

I see the concern, but don't really have any better ideas. `D` is at least right under the left hand for easy access.

Alternatively, I'm wondering if turning on the Wipe tool should enable the compare mode automatically, and turning off the Wipe tool should disable it. If so, perhaps we only need `W`?

> Is the opacity of the wipe tool right? I don't think I can really go much thinner and have it smooth, so adding some transparency does seem to help it not feel too obtrusive.

I don't think I would want it to be any weaker than it is now. My first impression was that maybe it was a bit too weak, but I'll defer to others. While on the subject I think I'd also be happier with a full circle for the rotation handle. But again, happy to defer.

> I've still got the commit where I stomp over the border colors in the stylesheet. Not sure what to do other than making this UI a special case.

Yeah, I'm afraid I don't have any great ideas here either. I'll have a little fiddle with it though...

A few other comments based on playing around with the feature :

1. The dark AB icon used to denote "None" kindof feels disabled as in you-can't-click-on-it. I tried pushing this to a lighter grey and I _think_ it feels better that way. Maybe it's just me - would be interesting to see what others think.
2. When I hover the AB button, the highlighted icon jumps in size which feels a little hokey. I'm not sure what's happening here but it'd be good to fix it. Maybe the auto-highlighting we do has a bug, or maybe the icons don't quite fit in the button, and changing them kicks Qt into some auto-scaling behaviour?
3. Would be good to fix the width of the `compare.view` dropdown to the same width as the `view` dropdown.
4. Perhaps it would be good to tie the WipeTool icon to the AB compare icon visually, by including an A and a B in it? Something vaguely like this?
![image](https://user-images.githubusercontent.com/1133871/186705639-76a3e5e9-aa7b-4917-a02d-2edf5255436b.png)
5. The solo channel modes are interacting badly with the compositing :
![image](https://user-images.githubusercontent.com/1133871/186707050-f1a600b4-65a7-46c2-b34f-9d07ecf8c7ae.png)
![image](https://user-images.githubusercontent.com/1133871/186707095-c7cdb671-9014-4fc1-bf30-70ea2f794c4e.png)
6. CPU mode exposure etc seems broken. Perhaps they're just not being applied to the second image?

danieldresser-ie commented:
Starting from the top, so first I'm replying to your overall comments:
> WidgetAlgo:joinEdges 
sure, that looks good.  I put it up as PR to merge first here: https://github.com/GafferHQ/gaffer/pull/4809

> Hotkeys
discussing with Murray, he wants one hotkey that turns on compare+wipe by default, but thought that having a hotkey for wipe could also be useful.  My new proposal: it's only possible to turn on wipe when comparison is active, and wipe is on by default.  I also suggested Q as a better hotkey than D - the letter isn't too meaningful anyway, and if we're not worried about overlapping with Scene tool keys, then Q is nicely placed key that is available, and having it next too, and before, W, is nice.  So the new interaction is:  Wipe tool is initially greyed out.  Press Q to enable both comparison and wipe.  You can now press W to disable/enable wipe while comparison is active.  When you toggle Comparison mode, you no longer see the wipe, but the wipe enabled/disabled state will be remembered when you turn it back on.  Murray seemed to think this was better than any of the other options we'd thought of, and I think it should be pretty nice.  Only thing is, if wipe is dependent on comparison, should it still be a tool on the left that is greyed out whenever comparison is inactive, or should it be controlled by an icon on the comparison toolbar at the top?

> styling of the wipe manipulator
Murray prefers half-circle to circle, and didn't mind the transparency.  I think I'll leave it as-is, we can always go a little bit more opaque later if users find images where the contrast is a problem ( I haven't really had any problems finding the manipulator unless I'm comparing high frequency noise to itself

> border colors in the stylesheet
Bit more complicated, second comment to follow on that.

> The dark AB icon used to denote "None" kindof feels disabled - I tried pushing this to a lighter grey and I think it feels better - sure I'm happy with that
  
> When I hover the AB button, the highlighted icon jumps in size which feels a little hokey
In _qtPixmapHighlighted, we render a rect defined by pixmapItem.boundingRect().bottomRight().  The doc for bottomRight states: `for historical reasons this function returns QPoint(left() + width() -1, top() + height() - 1)."`
This means it would make sense to add 1 to this result, but that made it worse.  Subtracting 1, on the other hand, appears to make everything line up perfectly.  I've added a commit to do this, maybe you can explain why it works.

> Would be good to fix the width of the compare.view dropdown to the same width as the view dropdown.

Done.  Does make sense to set them the same, but now that the header toolbar is getting a bit more crowded, do they need to be that large?  Would it make more sense to set them both to 125 pixels or something instead of 175?

> Perhaps it would be good to tie the WipeTool icon to the AB compare icon visually, by including an A and a B in it? Something vaguely like this?

Sure, that's reasonable.  I just increased the brightness a bit on the dim B, so that it's visible when the tool isn't active or hovered.

> The solo channel modes are interacting badly with the compositing :
Fixed

> CPU mode exposure etc seems broken. Perhaps they're just not being applied to the second image?
Addressed via hiding CPU mode, as discussed.

danieldresser-ie commented:
Re: the border colors for adjoined edges in the stylesheet.

My first premise: using transparency for adjoined edges has never really been fulfilling its intended function.  It would make sense if the edges were overlayed over the adjoined widgets background - but that's not how it's been working, it's showing through to whatever is behind the widget.  This has only worked acceptably so far because in the cases where we've actually used these colors, the background has been gray.

Based on this, my goal is just to choose opaque colors to use for adjoined edges which recreate the current appearance of adjoined edges.  I'm looking at the play/stop buttons on InteractiveArnoldRender.

After setting appropriate opaque colors, the appearance is pretty close, but the 2 pixels underneath the edge take the edge colors rather than blending into the bottom edge.  It looks like this can actually be resolved pretty well by using 1 pixel border radius instead of a 0 pixel border radius - it still goes all the way to the corner, but it causes the corner pixels to get a blended color, so that the corners look a bit more like they are actually shaded embossing.

From left to right:  [ Current Gaffer ] [ Opaque Colors ] [ Opaque Colors + 1 pixel radius ]

![joinEdgesOptions](https://user-images.githubusercontent.com/27969079/186792220-bc476a76-04e3-45e2-b208-94673bfdeaeb.png)

danieldresser-ie commented on include/GafferImageUI/ImageGadget.h:
```diff
@@ -144,11 +144,24 @@ class GAFFERIMAGEUI_API ImageGadget : public GafferUI::Gadget
 
 		Imath::V2f pixelAt( const IECore::LineSegment3f &lineInGadgetSpace ) const;
 
+		void setWipe( bool enabled, const Imath::V2f &position, const Imath::V2f &direction );
```
Switched to individual accessors.

johnhaddon replied:
It'd be very useful if you could link to the commit with the fix, to aid navigation for the reviewer. In this case, navigation is particularly hard because you've rebased and all the commits - including the originals - have moved below the line of the last review. I'd also recommend using `git commit --fixup` so that later squashing becomes trivial via `git rebase -i --autosquash`, and its doubly clear what is new and what's old. Trivial example here : https://github.com/GafferHQ/gaffer/pull/4807.

danieldresser-ie replied:
Switched to angle in 437b0a5

danieldresser-ie commented on include/GafferImageUI/WipeTool.h:
```diff
@@ -0,0 +1,83 @@
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSIMAGEI_WIPETOOL_H
```
Fixed.  And yes, we should switch to pragma once.  I vaguely recall that you don't want us to use pragma once anywhere until after we do a massive PR to switch all existing include guards?

danieldresser-ie commented on src/GafferImageUI/WipeTool.cpp:
```diff
@@ -0,0 +1,508 @@
+			Rotate
+		};
+
+		HandleSelect hoveredHandle( const ButtonEvent &event ) const
+		{
+			const V2f p = eventPosition( event );
+
+			float rsf = rasterScaleFactor();
+			if( ( p - m_pos ).dot( m_dir ) < rsf * g_handleThickness )
+			{
+				return HandleSelect::Translate;
+			}
+
+			return HandleSelect::Rotate;
+		}
```
Fixed.

danieldresser-ie replied:
Good catch - I had noticed this and tried to fix it, but I think it randomly worked when I tested - the fix wasn't doing anything.  Should be actually fixed now.

danieldresser-ie replied:
Fixed in 1c26fe2

danieldresser-ie commented on src/GafferImageUI/WipeTool.cpp:
```diff
@@ -0,0 +1,508 @@
+
+		void renderHandle( const Style *style, float rotateHandleSize, float thickness, float offset, const Color4f &color ) const
+		{
+			// Calculate a start and end point for the section of the infinite line defined by m_pos
+			// and m_dir that passes through the viewport.  Little bit complicated, but makes everything
+			// else simple.
```
My first pass was a very naive version of exactly that, which ran into weird numerical precision issues if the rotate handle was really far from the origin, at which point I just went for the fully complicated and maximally robust version.  But after implementing that, I realized that there was a simpler option that would work fine - do take the viewport into account, but only to find the point on the line which is closest to the viewport center, and base the drawing on that.  That is robust, and moderately simpler - since the complexity bothered you, I went ahead and implemented that version.

danieldresser-ie commented on python/GafferImageUI/ViewPlugValueWidget.py:
```diff
@@ -54,6 +54,8 @@ def __init__( self, plug, **kw ) :
 		self.__plugDirtiedConnection = self.getPlug().node().plugDirtiedSignal().connect( Gaffer.WeakMethod( self.__imagePlugDirtied ), scoped = True )
 		self._updateFromPlug()
 
+	def viewWidget( self ):
```
Yeah, this was only needed before the `joinEdges` fix.  Dropped.

danieldresser-ie commented on include/GafferImageUI/ImageView.h:
```diff
@@ -187,13 +191,14 @@ class GAFFERIMAGEUI_API ImageView : public GafferUI::View
 		bool m_lutGPU;
 
-		ImageGadgetPtr m_imageGadget;
+		ImageGadgetPtr m_imageGadget[2];
```
Fixed.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -93,6 +93,42 @@
+			"description",
+			"""
+			Enables a comparison mode to view two images at once - they can be composited under or over, or
+			viewed side by side.  Or replace mode just shows the front image, which is useful in combination
```
Fixed.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -93,6 +93,42 @@
+
+			"plugValueWidget:type", "GafferImageUI.ImageViewUI._ImageView_ViewPlugValueWidget",
+			"imageViewViewPlugWidget:ctrlModifier", True,
+			"toolbarLayout:width", 175,
+			"label", "",
```
Fixed.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1063,11 +1107,152 @@ def __stateChanged( self, imageGadget ) :
+	def __menuDefinition( self ) :
+
+		with self.getContext() :
+			compareMode = self.getPlug().getValue()
```
Oh, it was originally for that, but I deleted it after realizing that our menus currently don't show both an icon and a checkbox - do you want me to try and fix that?  Do you expect it to be much work?

johnhaddon replied:
I checked the Qt source (`qstylesheetstyle.cpp` if you're interested) and this looks like a no-go :

```
f (!mi.icon.isNull()) {
   // Draw icon
} else if (mi.menuHasCheckableItems) {
   // Draw checkbox
}
```

Which is a pity, because I really like the icons but I think the checkbox would be more useful to me.

danieldresser-ie replied:
![compareWipeUI](https://user-images.githubusercontent.com/27969079/188215690-332822b7-4352-487b-a521-80173dd1b9ab.png)

I was trying to figure out if it would be possible to show which one is selected with an icon, but it's tough when the icon is just16x16 pixels

danieldresser-ie replied:
Using the new icons was part of 1331762 - not sure if this is the best approach, but it feels tolerable to me?

danieldresser-ie commented on src/GafferImageUI/ImageView.cpp:
```diff
@@ -1111,12 +1111,13 @@ GAFFERIMAGEUI_API ImageView::ViewDescription<ImageView> ImageView::g_viewDescrip
 		m_soloChannelParameter( new IntData( -1 ) ),
 		m_lutGPU( true ),
-		m_imageGadget( new ImageGadget() ),
+		m_imageGadget( { new ImageGadget(), new ImageGadget() } ),
```
Apparently `m_imageGadgets{new ImageGadget(), new ImageGadget() }` is now valid.  Looks weird to me, but hopefully msvc likes it.

johnhaddon replied:
Looks like MSVC was happy with that, but found something else to complain about :

```
2022-08-26T02:08:43.1669208Z    Creating library lib\GafferImageUI.lib and object lib\GafferImageUI.exp
2022-08-26T02:08:43.1669965Z WipeTool.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl Iex_2_4::BaseExc::BaseExc(class Iex_2_4::BaseExc const &)" (__imp_??0BaseExc@Iex_2_4@@QEAA@AEBV01@@Z) referenced in function "public: __cdecl Iex_2_4::MathExc::MathExc(class Iex_2_4::MathExc const &)" (??0MathExc@Iex_2_4@@QEAA@AEBV01@@Z)
```

I think this should just be a case of adding `"Iex$OPENEXR_LIB_SUFFIX"` to the libs for GafferImageUI.

danieldresser-ie replied:
Fixed in d79f718

danieldresser-ie commented:
Made some pretty good progress on this feedback today, but didn't quite get through everything yet.  Uploaded where I'm currently at in case there's anything you want to take a look at.

Main things I could use comment on at the moment:
* does my new proposal for hotkeys ( in my very first reply today ) make sense
* how much work should I put into trying to hack Qt into showing both an icon and a checkmark ( it sounds pretty painful https://forum.qt.io/topic/32747/can-checkable-qactions-menu-items-with-icons-also-show-check-indicators ).  Given only the choice between the two, would it be better to leave the icon out of the menu?

johnhaddon reviewed (changes requested):
> My new proposal: it's only possible to turn on wipe when comparison is active, and wipe is on by default...

All sounds good to me, including the new hotkeys and everything else.

> Only thing is, if wipe is dependent on comparison, should it still be a tool on the left that is greyed out whenever comparison is inactive, or should it be controlled by an icon on the comparison toolbar at the top?

Having it in the comparison toolbar sounds much better - it would make the association between the two things much clearer. When presented with a greyed-out WipeTool icon, I think it'd be pretty non-obvious to folks that to enable it they need to go press a button in a different part of the UI. Having it in the toolbar means you only see the wipe icon when it's relevant, and it's one of the first things you see when enabling comparison mode for the first time.

> > styling of the wipe manipulator

> Murray prefers half-circle to circle, and didn't mind the transparency.  I think I'll leave it as-is

:+1: My first view of the wipe tool was on top of a checkerboard, which probably isn't representative.
  
> > When I hover the AB button, the highlighted icon jumps in size which feels a little hokey

> I've added a commit to do this

Great! That also fixes the annoying notch in the gamma icon that appears on highlight : ![image](https://user-images.githubusercontent.com/1133871/186885003-0df6b76b-5baa-403c-8528-08aab043f71c.png)

> maybe you can explain why it works.

I wish! The only thing I noticed is that the "for historical reasons" comment applies to QRect, but `boundRect()` is returning a QRectf. I've done enough deep-diving into the Qt source for one day, so I'm gonna be grateful for your fix and leave it unexplained. I think we probably need to apply the same fix in `_qtPixmapDisabled()` too though? 

> > Would be good to fix the width of the compare.view dropdown to the same width as the view dropdown.

> Done. Would it make more sense to set them both to 125 pixels or something instead of 175?

Makes sense to me.

johnhaddon commented on python/GafferUI/_StyleSheet.py:
```diff
@@ -113,14 +113,23 @@
 	"tintDarkerStrong" :    ( 0, 0, 0, 40 ),
 	"tintDarkerStronger" :    ( 0, 0, 0, 70 ),
 
+	# Borders within controls need to sit within the colors of the control, not
+	# the colors of the background.  There are situations where using
+	# transparency would show through to the background behind the control, not
+	# the background of the control, in which case it's better to use an opaque
+	# tint, relative to the standard grey we usually use for control backgrounds.
+	"internalTintLighter" :  ( 88, 88, 88 ),
+	"internalTintDarker" : ( 78, 78, 78 ),
```
> My first premise: using transparency for adjoined edges has never really been fulfilling its intended function. It would make sense if the edges were overlayed over the adjoined widgets background - but that's not how it's been working, it's showing through to whatever is behind the widget.

Makes sense.

> Based on this, my goal is just to choose opaque colors to use for adjoined edges which recreate the current appearance of adjoined edges. I'm looking at the play/stop buttons on InteractiveArnoldRender.

Also makes sense. I spent a bit of time googling to see if we could get Qt to draw the borders over the background, but I didn't come up with anything.

> After setting appropriate opaque colors, the appearance is pretty close, but the 2 pixels underneath the edge take the edge colors rather than blending into the bottom edge.

That's close enough for me.

I am a bit unhappy at the number of named colours in this stylesheet now though - it's hard to know what to use and when, and the use-transparency-for-tinting concept is creaking in yet another place. I have a few proposals there, none of them great :

1. Stick with introducing two new colours as you've done. But call them `backgroundDarkHighlightSubtle` and `backgroundLightHighlightSubtle` and put them with the original `backgroundDarkHighlight` and  $backgroundLightHighlight` colours which form the border edges for non-adjoining buttons. I hope by placing them with their associated values and using the same `Subtle` terminology as elsewhere might make this a tiny bit more understandable?
2. Don't modify the colours for adjoining buttons at all. This makes the join harder, and is less visually pleasing, but avoids yet more complexity in the stylesheet (picture below).
3. Embark on a simplification of the stylesheet based on an acceptance that tints don't work as well as we'd like, with the goal of reducing the complexity of the thing. Obviously this isn't an option for now.

![image](https://user-images.githubusercontent.com/1133871/186893876-54341153-beca-4ba0-973c-70afb46d4b5f.png)

johnhaddon replied:
Actually, thinking further, I don't like either the "Subtle" or "Internal" modifiers for these colours. The former doesn't give you any clue what the purpose of the colour is, and the latter does but deviates from the terminology we're already using. So I think we should use "Adjoined" instead, making it completely clear what it is for. In which case they would be `backgroundDarkHighlightAdjoined` and `backgroundLightHighlightAdjoined`. 

I don't know why the button uses `backgroundDarkHighlight` instead of `backgroundLightLowlight` when the documentation suggests that you use `<backgroundType><Highlight|Lowlight>` and `backgroundType` here would be `Light` according to the other colour assignments for the button. I guess it's just that it looks nicer that way...

johnhaddon replied:
> 3. Embark on a simplification of the stylesheet based on an acceptance that tints don't work as well as we'd like

Obviously we're not doing this now, but I think one principal for this might be that tinting should never be used for content widgets (buttons, text fields, tree views, other things you interact with). So those all have the same visual appearance regardless of nesting. And then tinting only gets used on backgrounds to indicate nesting - tabbed containers, frames and the like. Maybe that was the original idea anyway, and the buttons were just an anomaly?

danieldresser-ie replied:
I've gone with backgroundDarkHighlightAdjoined in b2ca2ad

I also went back to 0px corners, since it seems like the 1px ones might look weird at high DPI, and you said you were OK with how the 0px ones looked.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1063,11 +1107,152 @@ def __stateChanged( self, imageGadget ) :
+
+		node = self.getPlug().node()
+		inputImage = node._getPreprocessor()["in"]
+		viewNames = inputImage.viewNames()
```
Fixed in 8e4d75b

danieldresser-ie commented on src/GafferImageUI/ImageView.cpp:
```diff
@@ -1129,6 +1130,11 @@ ImageView::ImageView( const std::string &name )
+	PlugPtr compareParent = new Plug( "compare" );
+	addChild( compareParent );
+	compareParent->addChild( new StringPlug( "compareMode", Plug::In, "", Plug::Default & ~Plug::AcceptsInputs ) );
+	compareParent->addChild( new StringPlug( "compareView", Plug::In, "default", Plug::Default & ~Plug::AcceptsInputs ) );
```
Fixed in acd0fa3

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1185,16 +1181,17 @@ def _updateFromPlug( self ) :
+		if v:
+			GafferUI.WidgetAlgo.joinEdges( self.parent()[:], GafferUI.ListContainer.Orientation.Horizontal )
+		else:
+			GafferUI.WidgetAlgo.joinEdges( [ self ], GafferUI.ListContainer.Orientation.Horizontal )
```
Fixed in 2baefb2

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1185,16 +1181,17 @@ def _updateFromPlug( self ) :
-		self._repolish()
+		# If the mode is not empty, show the other widgets under the "compare" plug, and join
+		# all the widgets together
 		if self.parent():
```
Fixed in 5872fa1

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1055,11 +1107,149 @@ def __stateChanged( self, imageGadget ) :
+		m = IECore.MenuDefinition()
+		for name, value in [
+			( "None", "" ),
+			( "Replace", "replace" ),
```
Fixed in 5872fa1

danieldresser-ie commented:
OK, I think I've now completed a pass on feedback - this is badly in need of a rebase now, but I'll wait until after look at it.

Main remaining questions are UI related - do you like the new Wipe icon?  Is the indicator for the active option in the Mode menu enough?  Also, it's pretty weird that the Wipe toggle button gets focussed on after clicking it, but this appears to be a general Gaffer behaviour - I see the same thing clicking on the toggles on a CustomAttributes ... is there an easy way to turn this off?

johnhaddon commented:
> do you like the new Wipe icon?

Now it's next to the main AB icon, I think it's confusingly similar to it. I think dropping the AB and focussing on the appearance of the tool itself might be good. See my lame five minute job below for one possible direction...

![image](https://user-images.githubusercontent.com/1133871/188477827-a20a90a2-2003-43f0-9ff2-9bc93e57779f.png)

> Is the indicator for the active option in the Mode menu enough?

I think so. I'd quite like to see a version with white brackets rather than blue, and a bit more spacing inside the brackets if possible. I also wonder if we could hack this in Menu.py, so that we automatically generate a wider icon with a check mark inside it when we realise we want a checkmark as well as an icon. Might be a bit out of scope for this PR though...

> Also, it's pretty weird that the Wipe toggle button gets focussed on after clicking it, 

See https://github.com/GafferHQ/gaffer/commit/19f358133966462496e0f258ea349c90d299b48c for a fairly easy way of turning it off. I haven't tested this particularly thoroughly, but it seems reasonable that we might just do this across the board?

One last thing : the <kbd>Q</kbd> shortcut is breaking the <kbd>Ctrl</kbd>+<kbd>Q</kbd> shortcut for Quit. We need to check that there are no modifiers before accepting the event.

danieldresser-ie commented:
>  I think dropping the AB and focussing on the appearance of the tool itself might be good.

I've taken a swing at this.

> I'd quite like to see a version with white brackets rather than blue, and a bit more spacing inside the brackets if possible.

I've switched to white brackets.  Unfortunately more space is not easy to achieve.  Searching around online, it sounds like it is possible, but is a real headache: https://forum.qt.io/topic/63483/qmenu-icon-size
Maybe if this comes up more frequently we can implement the "automatic append a checkmark to the icon" hack and figure that out?

> I haven't tested this particularly thoroughly, but it seems reasonable that we might just do this across the board?

Ivan felt that he would expect the behaviour of clicking on buttons setting focus in general ( I'm dubious of that ... ).  Since it's not an unambiguous win, I've gone with just setting it on the Wipe button for now, since that's the only place it really feels like a problem.

> One last thing : the Q shortcut is breaking the Ctrl+Q shortcut

Fixed.

The only other thing I noticed while prepping for merge:  there's a theoretical performance regression:  while viewing a solo'ed channel, and not using comparison, you don't technically need to evaluate the alpha, but we now always provide the alpha in case it is needed, which does mean evaluating some tiles for another channel.  I don't think this special case is particularly concerning - if we were going to address this, there are cases where you aren't soloing a channel where providing the alpha is unnecessary, and it would probably be more useful to address those.

Otherwise, I think this is good to go - it's all rebased and squashed, and I've taken a pass on the Changes.md ( though I haven't split the changes up into individual commits to Changes, let me know if I should do that tomorrow ).

johnhaddon commented:
I've got some Qt/UI nitpicks still, but I think perhaps the most pragmatic thing is to merge anyway and then address/debate those afterwards? I'll leave that up to you - if you want to merge, then please do.

Here goes :

- The wipe button seems to have lost the light-to-dark gradient that gives buttons their slightly raised feel. It makes it feel like it's maybe pushed in, when its not.
- The focus hack on the one button seems worse than not having it. Now when I click on the wipe button, focus seems to randomly jump to a button either side of it, which then gets highlighted. I think we're better going all-or-nothing here, and I think I'm in the "all" camp, so we might need to debate with Ivan.
- I don't really "get" the new wipe button. Mainly because our other on/off toggles get uniformly brighter when on, but this gets darker in places, I think. But possibly also because the circle doesn't really correspond to the wipe tool's half circle? Probably a matter of taste - @murraystevenson tie-breaker perhaps? 
- I've lost track of the discussion on it, so perhaps I'm at fault here, but I think the 1px radius is actually mildly better for adjoined buttons, even on my high DPI screen.
- This hasn't happened in this PR, but the pixel inspector overlay has lots its alignment somewhere on `main`, and is now aligned to the right rather than the left where it should be.
- I tried a "use the icon when off and use the checkbox when on" gambit for the menu, and it didn't look _terrible_. Perhaps worth considering?

danieldresser-ie commented:
> The wipe button seems to have lost the light-to-dark gradient that gives buttons their slightly raised feel. It makes it feel like it's maybe pushed in, when its not.
As far as I can tell, it hadn't lost the gradient, but menu buttons have a 20% brighter gradient than other buttons, so it looks darker next to menu buttons.  I've now made the gafferThinButton style match the brighter gradient of menu buttons.

> The focus hack on the one button seems worse than not having it.
The behaviour you're seeing doesn't seem to match what I'm seeing ... it definitely feels much worse here without it, but I've removed it for now.  I am 100% in favour of removing it from all buttons, I just didn't want to make this PR depend on it.

> I don't really "get" the new wipe button.
Oh, oops, I didn't get your suggestion, and thought the half circle was supposed to represent a sample image being cut in half, rather than being part of the handle.  I've had a quick go at just imitating your suggestion now that I get what you meant.

> I've lost track of the discussion on it, so perhaps I'm at fault here, but I think the 1px radius is actually mildly better for adjoined buttons, even on my high DPI screen.
Cool, I prefer it, I think I thought you disliked it just because you didn't comment on it.  Set back to 1px.

> This hasn't happened in this PR, but the pixel inspector overlay has lots its alignment somewhere on main
Since unrelated to this PR, haven't addressed yet.

> I tried a "use the icon when off and use the checkbox when on" gambit for the menu, and it didn't look terrible. Perhaps worth considering?

Much simpler, and seems reasonable.  I've done that.

johnhaddon commented:
Thanks Daniel - I'll let CI run through and then fix the inevitable Changes.md conflict and merge. 

>> The focus hack on the one button seems worse than not having it.

> The behaviour you're seeing doesn't seem to match what I'm seeing

Here's what I see - when I click the wipe button, the focus jumps to the mode button (or the view button if that's what I clicked last) :

![Peek 2022-09-08 09-48](https://user-images.githubusercontent.com/1133871/189078706-7b860f01-1529-4ab1-8cc5-d39539106237.gif)

> I am 100% in favour of removing it from all buttons, I just didn't want to make this PR depend on it.

Let's follow up on this right away then. @ivanimanishi, do you have important use cases for "click to focus button, then use keyboard to press it repeatedly?". Or is there another consideration we're missing?

ivanimanishi commented:
> Let's follow up on this right away then. @ivanimanishi, do you have important use cases for "click to focus button, then use keyboard to press it repeatedly?". Or is there another consideration we're missing?

No. As I explained to Daniel, it was just an initial general expectation that when one interacts with a widget, it grabs the focus, and the fact that Qt's default behaviour seems to be doing just that, even for Buttons. But I couldn't think of any real case scenario for Buttons where I definitely wanted that to be the case, so I have no reservations about making the change a generic one for Buttons in gaffer.
